designs/sam_build_cmd.rst

```diff
@@ -0,0 +1,440 @@
++-------------+---------------------------------------------------------------------------------------+
+
+
+Custom Build Commands
+~~~~~~~~~~~~~~~~~~~~~
+
+Users can provide custom build command for each runtime via ``.samrc`` that will be run instead of the default build
+action provided by SAM CLI. SAM CLI will invoke this command for every resource that uses this runtime. This can be any
+arbitrary program that is accessible (via ``PATH`` resolution). The program will be invoked by setting three
+environment variables:
+
+#. ``__SOURCE_DIR`` - Folder where the source files are located
+#. ``__BUILD_DIR`` - Folder where the built artifacts should be written to
+#. ``__RUNTIME`` - AWS Lambda function runtime
+#. ``__TEMPLATE`` - Path to SAM Template
+#. Current Working Directory - Same directory where ``sam build`` command was invoked from
+
+This command is supposed to return an exit code of ``zero`` to indicate success. Non-zero exit codes indicate failure.
+
+Building Native Binaries
+~~~~~~~~~~~~~~~~~~~~~~~~
+To build native binaries, we need to run on an architecture and operating system that is similar to AWS Lambda. We use
+the Docker containers provided by `Docker Lambda <https://github.com/lambci/docker-lambda>`_ project to run the same
+set of commands described above on this container. We will mount source code folder and build folder into the container
+so the commands have access to necessary files.
+
+If the user specified a custom build command, we will attempt to run the command as-is within the Docker Container.
+But it is possible that the command might not be available inside the container.
+
+``.samrc`` Changes
+------------------
+*Explain the new configuration entries, if any, you want to add to .samrc*
+
+We will add a new section to ``.samrc`` where customers can provide custom build actions. This section will look like:
+
+.. code-block:: json
+
+    {
+        "build": {
+            "actions": {
+                "java8": "gradle build",
+                "dotnetcore2.1": "./build.sh"
+            }
+        }
+    }
+
+Security
+--------
+
+*Tip: How does this change impact security? Answer the following questions to help answer this question better:*
+
+**What new dependencies (libraries/cli) does this change require?**
+
+**What other Docker container images are you using?**
+
+**Are you creating a new HTTP endpoint? If so explain how it will be created & used**
+
+**Are you connecting to a remote API? If so explain how is this connection secured**
+
+**Are you reading/writing to a temporary folder? If so, what is this used for and when do you clean up?**
+
+**How do you validate new .samrc configuration?**
+
+
+Documentation Changes
+---------------------
+TBD
+
+Open Questions
+--------------
+
+#. Should we support ``artifacts.json`` now to be future-proof?
+#. Should we create the default ``build`` folder within a ``.sam`` folder inside the project to provide a home for
```

Comment:
+1 to `.sam` folder as this provides room for growth (i.e. transform, build, dev plugins, etc)

Reply:
Agreed. My concern around `.sam` folder for this revision is inventing too many features at once and ending up with a lot of debt. Can we live without `.sam` folder for this release and add it later?

Reply:
A way to weigh the now vs later:
Is this `build` folder something we will encourage customers to edit or mutate? If so, moving this later becomes more difficult, as we have to be concerned with breaking/change customer behavior.

On the same note: Should `build` be a hidden folder?

Reply:
I vote for having `.sam` but no `artifacts.json`. `.sam` folder is easier to relate to and can easily be added to `.gitignore` so you don't have to add both `built-template.yaml` + `build` folder and whatever else comes afterward.

For now, I see that as a way to keep things tidy that will avoid polluting their workspace. 

Hope that makes sense

Reply:
@jfuss  No customers are not expected to mutate the build folder, but it is not something we can control.

@heitorlessa I agree with your reasoning. It does standardize a few things. But it also adds new surface area for things we need to maintain. We need to think thru questions like where does the folder exist, can it exist in multiple places (project root & user home), can other tools consume the contents of this folder, how does sam cli operate in absence of this folder, how do CI/CD systems use this folder to extra artifacts etc. I am just not sure if this proposal is the right place to address these questions. It certainly adds another reason why we need this folder.

Can you create a Github Issue instead so we can track this work for future? It will also help get feedback from community

Reply:
Happy to create a new issue for that but I think we should discuss here - Here's the answers to these questions for a MVP:

#### Can it exist in multiple places (project root & user home)?

No, just project root. `.sam` folder will contain build artifacts and it's project specific; moving to user home brings unwanted side effects (disk space and logic complexity).

#### Q: Can other tools consume the contents of this folder?

`.sam` folder contains read-only build artifacts so any tool built to work with sam can read from that.

#### Q: How does sam cli operate in absence of this folder?

We should create it in the absence of it - That provides a `Just Works(tm)` experience. Here's a sample scenario for this:

##### No explicit sam build from the user

**UX**

This is the ideal scenario assuming `.sam` folder contains artifacts, packaged-template and other commands are aware of `.sam` and `.samrc`:

* Initialize boilerplate: `sam init --name my-project && cd my-project`
* Deploy: `sam deploy`

SAM application will be deployed seamless and the entire process logged. One could introduce `sam local` before `sam deploy` and `sam build` will run seamlessly if `.sam` folder is absent.
Power users can then modify the convention and opinions if they wanted to.

**Logic**

* `sam deploy` checks `.sam` absence and run `sam build` for the runtime chosen in `template.yaml`
* `sam deploy` then run `sam package`
* `sam package` creates a S3 bucket based on project name (`.samrc`) if one doesn't exist
* `sam deploy`  ultimately deploys the application
* `.samrc` is instrumental in allowing the user to add his/her opinions to change this flow (e.g. custom build, custom stack name, custom s3 bucket name)

**Pros**

1. Seamless experience and just works without having to know `built-template.yaml`, `packaged-template.yaml`, etc.
2. New users don't have to learn upfront how Lambda packaging works and embed `sam local` within their package managers
    - `npm run build -> sam build`
    - `npm run deploy -> sam deploy`
    - `make build -> sam build`
    - `make deploy -> sam deploy`
3. Power users can rely on `.sam` and build wrappers/tools until `sam package/deploy` changes its logic

**Cons**

1. This can only happen post-MVP as `sam local`, `sam package`, and `sam deploy` would need to check for `.sam`

#### Q: How do CI/CD systems use this folder to extra artifacts etc.

Extra artifacts is a responsibility of the CI system and SAM only takes into account the application artifact so CloudFormation can deploy it afterwards - But perhaps I didn't understand the question.

Here's an example using CodeBuild where CodePipeline would pass that as an Input Artifact to CloudFormation:

```yaml
phases:
  install:
    commands:
      - echo "[Security phase]"
      - pipenv check # run some CVE checks on deps
  pre_build:
    commands:
      - echo "[Pre-Build phase] Nothing to be done as SAM build now handles that for me ;)"
  build:
    commands:
      - echo "[Build phase] Starting SAM build `date` in `pwd`"
      - sam build
      - echo "[Build phase] SAM build completed on `date`"
  post_build:
    commands:
      - echo "[Post-build] Uploading artifacts to S3 `date`"
      - sam package

artifacts:
  files:
    - .sam/template-packaged.yaml
  discard-paths: yes
```

By supporting `.sam` folder and only storing `artifacts` this can set the path for a fast follow up on `sam package` and `sam deploy` afterwards.

Reply:
@heitorlessa  thanks for the detailed thoughts. It is very encouraging to see this mature. I created https://github.com/awslabs/aws-sam-cli/issues/748 to follow up on a separate thread. This needs proper design to make sure we are making the right decisions. I don't want to haste into it.